specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1098 @@
+
+- __Invalid state transition__: Any malicious majority of a subnet can also push invalid state transitions in a valid block. Malicious validators can include messages in a valid block that trigger unallowed state changes. Invalid state transitions are a general misbehavior that may impact HC-specific processes like the commitment of checkpoints or the result of atomic executions.
+In order to report an invalid state transition in a subnet, users need to submit to the parent SCA the chain of blocks from the closest valid committed checkpoint to the block where the invalid state transition happens. The parent SCA replays every state change in the reported chain and verifies that, indeed, an invalid state transition happened in the reported block. If the misbehavior is reported successfully, the collateral for the subnet is correspondingly slashed.
+Thus, if an invalid state transition happens at block `n`, and the latest committed checkpoint was at block `n-x`, a user looking to report the misbheavior needes to submit to the SCA of the parent the chain of `x` blocks from `n-x` to `n` so the state can be replayed and the invalid state transition checked.
+>Note: Along with the checkpoint at `n-x` a snapshot of the state at `n-x` may also need to be provided to enable the whole state of the subnet (for every contract) to be replayed.
+
+## Cross-net routing gas price
+> WIP. This is just a placeholder for this section. The content included is not final. Follow all design updates [here](https://hackmd.io/HpxNIaacTn2_t6jaDs9RcQ)
+
+Cross-net messages from a source subnet, `sn_src`, to some other subnet in the hierarchy,`sn_dest`, need to be provided with enough gas for the message to be executed and routed successfully by every subnet in the path `Path(sn_src, sn_dest)`. Cross-net messages within a subnet are treated like any other message, and their execution triggers some state change that costs a certain amount of gas in the subnet, `gas_cost_sn`. The main difference for cross-net messages is that this execution and state change may translate into the propagation of the message to the next subnet in the path. However, for a user to be able to provide the message with enough gas, it needs to have a sense of the cost of execution in each subnet in the path.
+
+To improve the UX and make cross-net gas costs more predictable, each subnet publishes basic parameters of their gas model at spawning time: 
+- `gas_model_type` (if based on miner tipping, EIP1559-like, etc.)
+- `curr_base_fee` (if any)
+- etc.
+
+With this information, cross-net message will include an array of `gas_limit` amounts that the message is willing to allocate to each subnet in the hop. Subnets in the path won't be able to charge over the specified `gas_limit` for the subnet. When a cross-net reaches a subnet for which `gas_limit` is insufficient, it fails and an error message is propagated back to the source, along with the unspent gas.
```

Comment:
Do you think it's possible to charge a flat fee for pass-through messages, ones that only need their meta-data appended to the top/bottom message list? I forget if they need message resolution, I think they do. But they don't get executed as smart contracts. 

Or are fees expected to fluctuate by demand? Perhaps have a flat rate in gas, but cost various amount of FIL based on prevalent gas prices? 

I can see how it would be a nightmare for users to keep track of where their message bounced, or even provision up front, because the `Value` is in FIL, while the `gas_limit` I assume is not, so in the end the final value delivered may be less than they intended. 

They might also expect that the network should delay the inclusion of their messages to a time when fees are more reasonable.

Reply:
I suppose if we add price to the fees, then delaying is out of the question as it would necessarily have to hold up all other cross net messages as well, because of the nonces, to avoid out-of-order delivery.

Reply:
Just noting that the mechanism to signal errors and return gas is not clearly reflected in the SCA interface (or SA). If it's some kind of cross message that source actors will receive, it should perhaps be specified so those actors know what they have to handle.

Reply:
> I can see how it would be a nightmare for users to keep track of where their message bounced, or even provision up front, because the Value is in FIL, while the gas_limit I assume is not, so in the end the final value delivered may be less than they intended.

Definitely, this really concerns me. We've had a lot of back-and-forth with the CEL team to see what is the best way to have this designed. As mentioned above, I think this deserves its own thread, so let me open a new PR to discuss this design. I'll ping you there.